botfront/imports/lib/importers/validateDomain.js

```diff
@@ -0,0 +1,387 @@
+                // the existing one are put in first so during deduplication they will be kept
+                const newResp = deduplicate([...existingResps, ...currentResp], 'lang');
+                toInsert = { ...toInsert, [respKey]: newResp };
+            } else {
+                toInsert = { ...toInsert, [respKey]: currentResp };
+            }
+        });
+        // "toInsert" is after "all", because "toInsert" might contain an updated version of a respone in "all"
+        return { ...all, ...toInsert };
+    }, {});
+    const allSlots = filesToProcess.reduce((all, { slots = {} }) => ({ ...slots, ...all }), {});
+    const allForms = filesToProcess.reduce((all, { forms = {} }) => ({ ...forms, ...all }), {});
+    const allAction = filesToProcess.reduce((all, { actions = [] }) => [...all, ...actions], []);
+    const mergedActions = deduplicateArray(allAction); // we are not using a set to deduplicate to keep the order of the actions
+    return {
+        slots: allSlots,
+        responses: allResponses,
+        forms: allForms,
+        actions: mergedActions,
+    };
+};
+
+const validateADomain = (
+    file,
+    {
+        defaultDomain = {}, projectLanguages = [], actionsFromFragments = [], fallbackLang,
+    },
+    // we validate domain and default domain with the same function
+    // isDefaultDomain allow us to get the domain in rasaformat
+    // and also trigget specific warning linked with default domain
+    isDefaultDomain = false,
+) => {
+    const { rawText } = file;
+    let domain;
+    try {
+        domain = safeLoad(rawText);
+    } catch (e) {
+        return {
+            ...file,
+            errors: [...(file?.errors || []), 'Not valid yaml'],
+        };
+    }
+    const {
+        slots: slotsFromFile = {},
+        templates: legacyResponsesFromFile = {},
+        responses: modernResponsesFromFile = {},
+        forms: formsFromFile = {},
+        actions: actionsFromFile = [],
+    } = domain;
+    const {
+        slots: defaultSlots = {},
+        responses: defaultResponses = {},
+        forms: defaultForms = {},
+    } = defaultDomain;
+    const responsesFromFile = {
+        ...(legacyResponsesFromFile || {}),
+        ...(modernResponsesFromFile || {}),
+    };
+    // get forms MIGHT NEED TO UPDATE WITH RASA 2
+    const bfForms = 'bf_forms' in (slotsFromFile || {}) ? slotsFromFile.bf_forms.initial_value : [];
+    // do not import slots that are in current default domain or are programmatically generated
+    [...Object.keys(defaultSlots), ...INTERNAL_SLOTS].forEach((k) => {
+        delete slotsFromFile[k];
+    });
+    // do not import responses that are in current default domain
+    Object.keys(defaultResponses).forEach((k) => {
+        delete responsesFromFile[k];
+    });
+    // do not import forms that are in current default domain
+    Object.keys(defaultForms).forEach((k) => {
+        delete formsFromFile[k];
+    });
+
```

Comment:
I started to make the change but actually do we want that ? since form and action will end up in the default domain anyway. I think it should be the other way around, e.g: If i import a form that has the same name as one in the default domain, I would expect the form that being imported to overwrite the one in the default domain, as it might be an update of the said form.

Reply:
I agree, so we make it so that forms and actions overwrite the ones with same name ?